clang/lib/CodeGen/CGHLSLRuntime.cpp

```diff
@@ -489,3 +494,100 @@ void CGHLSLRuntime::generateGlobalCtorDtorCalls() {
       GV->eraseFromParent();
   }
 }
+
+// Returns handle type of a resource, if the type is a resource
+// or an array of resources
+static const HLSLAttributedResourceType *findHandleTypeOnResource(QualType QT) {
+  // If the type is a resource class, the first field must
+  // be the resource handle of type HLSLAttributedResourceType
+  const clang::Type *Ty = QT->getUnqualifiedDesugaredType();
+  if (RecordDecl *RD = Ty->getAsCXXRecordDecl()) {
+    if (!RD->fields().empty()) {
+      const auto &FirstFD = RD->fields().begin();
+      return dyn_cast<HLSLAttributedResourceType>(
+          FirstFD->getType().getTypePtr());
+    }
+  }
+  return nullptr;
+}
```

Comment:
This overlaps with the similar static function in #111203 - how do we plan to share logic here?

Reply:
It would be nice to share this logic, but Sema and CodeGen currently do not have any dependencies between them. Maybe a static method on `HLSLAttributedResourceType`?

Reply:
I think that could make sense. We may even need something there that can walk arrays and/or UDTs to find the resource type as we fill out the TODOs here.